### PR TITLE
Compact run buffer slice values

### LINKED ISSUE #10017: RunArray::slice() should align run_ends and values with the logical slice
## `RunArray::slice()` should trim the underlying values array

`RunArray::slice()` currently tracks slicing logically without adjusting the underlying values array, so a sliced `RunArray` still references the full original values. This has performance implications for patterns that operate on `array.values()`.

### Example

```rust
let run = Int32Array::from(vec![3, 6, 9]);
let values = Int32Array::from(vec![1, 2, 3]);
// [1, 1, 1, 2, 2, 2, 3, 3, 3]
let array = RunArray::try_new(&run, &values).unwrap();

// Logically [2, 2, 2], but still references the original arrays:
// {1, 1, 1, [2, 2, 2], 3, 3, 3}
let array_sliced = array.slice(3, 3);
```

Ideally `array_sliced` would be `{ run_ends: [3], values: [2] }`, but the [current implementation of `RunArray::slice()`](https://github.com/apache/arrow-rs/blob/8acab7b5371470deff6c211899295d2bb3030dfc/arrow-array/src/array/run_array.rs#L344) preserves the full values array.

### Why it matters

Patterns like the following do extra work on values outside the slice:

```rust
let values = date_part(array.values(), part)?;
let new_array = array.with_values(values);
```

Not a correctness bug, but worth avoiding.

### Regression checks

Each call site of `RunArray::slice()` needs to be reviewed to ensure this change doesn't introduce breaking behavior.

### Reference

[[Original comment thread](https://github.com/apache/arrow-rs/pull/9959#discussion_r3270678608)](https://github.com/apache/arrow-rs/pull/9959#discussion_r3270678608)

### PR DESCRIPTION
**Pull request by Jefffrey**
# Which issue does this PR close?

- Closes #10017

# Rationale for this change

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: arrow-buffer/src/buffer/run.rs [hunk 1]
```diff
@@ -277,9 +277,32 @@ where
             logical_offset.saturating_add(logical_length) <= self.logical_length,
             "the length + offset of the sliced RunEndBuffer cannot exceed the existing length"
         );
-        Self {
+        let logical_offset = self.logical_offset + logical_offset;
```
Comment (Jefffrey) [inline arrow-buffer/src/buffer/run.rs:280]: cc @Rich-T-kid this is what i was thinking on my previous comments on your PR

essentially we can still reduce how many physical values we hold on to, without needing to readjust all of the values (still keeping it zero-copy)

we'd need to extend it to runarray too since need to ensure the values array is similarly compacted (might need a bit of plumbing to make it work) and as noted in the issue callsites may need to be adjusted

lemme know your thoughts

**Comment by Rich-T-kid** [comment]
will take a look this afternoon!

File: arrow-buffer/src/buffer/run.rs [hunk 2]
```diff
@@ -421,4 +432,53 @@ mod tests {
...
+        let slice = buffer.slice(0, 0);
+        assert_eq!(slice.len(), 0);
+        assert_eq!(slice.offset(), 0);
+        assert_eq!(slice.values(), &[4]);
+
+        // compact start only
+        // [B, B, B, B, B, B, {B, B, C, C, C, C}]
+        let slice = buffer.slice(6, 6);
+        assert_eq!(slice.len(), 6);
+        assert_eq!(slice.offset(), 6);
```
Comment (Jefffrey) [inline arrow-buffer/src/buffer/run.rs:460]: i realized my initial version of the PR incorrectly adjusted the offset when that wasnt needed; by removing the first physical value its as if the first run was consumed by the following run since we always list run lengths; that is

```
[4, 8] -> AAAABBBB
[8] -> {BBBB}BBBB
```

- where `{BBBB}` used to be a separate run but is now folded into the first run; this shouldnt matter as our logical offset points beyond it anyway

**Comment by Rich-T-kid** [comment]
Going to review this PR this week 👍

**Review by Rich-T-kid — CHANGES_REQUESTED** [review]
@Jefffrey I took a look an im a bit confused. Im going to go through `RunEndbuffer` to refresh my knowledge. If you get a chance could you add a test similar to what we found in https://github.com/apache/arrow-rs/pull/9959#discussion_r3270678608. This way we validate that the slice gives us what we expect.

will try to give a more insightful review tomorrow

**Comment by Jefffrey** [comment]
> @Jefffrey I took a look an im a bit confused. Im going to go through `RunEndbuffer` to refresh my knowledge. If you get a chance could you add a test similar to what we found in [#9959 (comment)](https://github.com/apache/arrow-rs/pull/9959#discussion_r3270678608). This way we validate that the slice gives us what we expect.
> 
> will try to give a more insightful review tomorrow

ill look into this; i didnt touch` RunArray` itself since it needs a bit of plumbing to make work but will try get around to it

File: arrow-array/src/array/run_array.rs [hunk 3]
```diff
@@ -1364,33 +1397,32 @@ mod tests {
 
     #[test]
     fn test_run_array_values_slice() {
```
Comment (Jefffrey) [inline arrow-array/src/array/run_array.rs:1399]: can reference this test to see how the physical values change underneath as we slice

**Review by Jefffrey — COMMENTED** [review]
another thought i had is if we enforce this strictly (e.g. in all construction & validation methods) it can lead to simpler code; for example, `values_slice()` would become deprecated since it would be the same as `values()`

https://github.com/apache/arrow-rs/blob/707872fbd58520ea09d0463ea5264eb96cc55e30/arrow-array/src/array/run_array.rs#L269-L280

and `get_start_physical_index()`/`get_end_physical_index()` can be simplified to always getting `0`/`len - 1` if we know the run ends buffer strictly keeps only needed physical values, essentially skipping the binary search

https://github.com/apache/arrow-rs/blob/707872fbd58520ea09d0463ea5264eb96cc55e30/arrow-buffer/src/buffer/run.rs#L242-L267

File: arrow-array/src/array/run_array.rs [hunk 4]
```diff
@@ -345,10 +345,50 @@ impl<R: RunEndIndexType> RunArray<R> {
...
+        let values = self.values.slice(start, end - start + 1);
+        // SAFETY: existing values are valid, and values referenced by the logical
+        //         offset are preserved
+        let run_ends = unsafe {
+            RunEndBuffer::new_unchecked(
+                tmp.inner().slice(start, end - start + 1),
+                tmp.offset(),
+                tmp.len(),
+            )
+        };
```
Comment (Rich-T-kid) [inline arrow-array/src/array/run_array.rs:386]: okay makes sense to me

File: arrow-array/src/array/run_array.rs [hunk 5]
```diff
@@ -1234,15 +1274,11 @@ mod tests {
             // test for offset = 0 and slice length = slice_len
             // slice the input array using which the run array was built.
             let sliced_input_array = &input_array[0..slice_len];
-
-            // slice the run array
-            let sliced_run_array: RunArray<Int16Type> =
-                run_array.slice(0, slice_len).into_data().into();
-
-            // Get physical indices.
+            let sliced_run_array = run_array.slice(0, slice_len);
```
Comment (Rich-T-kid) [inline arrow-array/src/array/run_array.rs:1277]: looks alot cleaner

**Review by Rich-T-kid — COMMENTED** [review]
Sorry it took me so long to take another look at this.


This mostly looks good to me, its still zero copy as the `.slice()` method contract promises.

mabey we could add a couple extra test that exercise edge cases (0 size, empty,invalid bound..ect) + a test that goes through the flow that started this issue/PR (https://github.com/apache/arrow-rs/pull/9959#discussion_r3270678608)
